cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -122,22 +122,24 @@ void put(String key, Value value, Scope scope) {
     Preconditions.checkNotNull(value.toString(), String.format("Null value provided for the key '%s'.", key));
     Preconditions.checkState(nodeName != null, "nodeName cannot be null.");
 
-    List<NodeValue> nodeValueList = tokenValueMap.get(scope).get(key);
-    if (nodeValueList == null) {
-      nodeValueList = Lists.newArrayList();
-      tokenValueMap.get(scope).put(key, nodeValueList);
-    }
+    synchronized (tokenValueMap) {
```

Comment:
Updated the code to use the private static member variable of the `BasicWorkflowToken` for synchronization. Let me know if this is not what you meant.